postgresql/src/main/scala/org/scalarelational/postgresql/PostgreSQLDatastore.scala

```diff
@@ -0,0 +1,145 @@
+ */
+
+sealed trait PGSsl
+case class PGUseSsl(sslfactory: Option[String] = None, sslfactoryarg: Option[String] = None ) extends PGSsl
+case class PGNoSsl() extends PGSsl
+
+case class PGConfig(host: String,
+                            schema: String,
+                            user: String,
+                            password: String,
+                            port: Int = 5432,
+                            useSsl: PGSsl = PGNoSsl()
+                           )
+
+abstract class PostgreSQLDatastore private() extends SQLDatastore with Logging with SQLLogging {
+  protected def this(pgConfig: PGConfig) = {
+    this()
+    sqlLogLevel := Level.Warn
+    config := pgConfig
+  }
+
+  override def supportsMerge = false
+
+  protected def this(dataSource: DataSource) = {
+    this()
+    dataSourceProperty := dataSource
+  }
+
+  Class.forName("org.postgresql.Driver")
+
+  val config = Property[PGConfig]()
+
+  init()
+
+  protected def init() = {
+    config.change.on {
+      case evt => updateDataSource() // Update the data source if the mode changes
+    }
+  }
+
+  def updateDataSource() = {
+    dispose() // Make sure to shut down the previous DataSource if possible
+
+    val source: PGSimpleDataSource = new PGSimpleDataSource()
+    source.setPortNumber(config().port)
+    source.setServerName(config().host)
+    source.setDatabaseName(config().schema)
+    source.setUser(config().user)
+    source.setPassword(config().password)
+    config().useSsl match {
+      case PGUseSsl(sslfactory, sslfactoryarg) =>
+        source.setSsl(true)
+        sslfactory.foreach {f => source.setSslfactory(f)}
+        sslfactoryarg.foreach {f => source.setSslFactoryArg(f)}
+      case PGNoSsl() =>
+    }
+    dataSourceProperty := source
+  }
+
+  override protected def columnPropertiesSQL(container: ColumnPropertyContainer): List[String] = {
+    val b = ListBuffer.empty[String]
+    if (!container.isOptional && !container.has(Polymorphic)) {
+      b.append("NOT NULL")
+    }
+    if (container.has(Unique)) {
+      b.append("UNIQUE")
+    }
+    b.toList
+  }
+
+  override protected def columnSQLType(create: CreateColumn[_]) =
+    if (create.has(AutoIncrement)) {
+      "SERIAL"
+    } else if (create.dataType == doubleDataType) {
+      "DOUBLE PRECISION"
+//    } else if (create.dataType == byteArrayDataType) { //Not sure if this isn't tested or works without?
+//      "BYTEA"
+    } else if (create.dataType.isInstanceOf[ObjectSerializationConverter[_]]) {
+      "BYTEA"
+    } else if (create.dataType == blobDataType) {
+      "BYTEA"
+    } else {
+      super.columnSQLType(create)
+    }
+
+  override def jdbcTables = {
+    val s = session
+    val meta = s.connection.getMetaData
+    val results = meta.getTables(null, "public", "%", null)
+    try {
+      new ResultSetIterator(results).map(_.getString("TABLE_NAME")).toSet
+    } finally {
+      results.close()
+    }
+  }
+
+  override def jdbcColumns(tableName: String) = {
```

Comment:
This is identical to the super-method with the only difference that ``public`` is lowercase. Let's avoid the duplication by providing overrideable functions for all schema patterns. Then we could get rid of ``jdbcTables`` and ``doesTableExist`` too.